src/GNDStk/Node/src/add-child.hpp

```diff
@@ -0,0 +1,283 @@
+}
+
+
+
+// -----------------------------------------------------------------------------
+// child_t<void>, *
+// Guaranteed to add something
+// Returns reference to added node
+// -----------------------------------------------------------------------------
+
+// Similar in principle to its meta_t counterpart. Bounces to one of the above
+// add() functions; and like those, doesn't allow std::optional.
+
+// child_t<void>, plain
+// child_t<void>, defaulted
+template<
+   class T = Node, allow ALLOW, class FILTER,
+   class = typename std::enable_if<std::is_constructible<Node,T>::value>::type,
+   class = typename std::enable_if<!detail::is_optional<T>::value>::type
+>
+Node &add(
+   const child_t<void,ALLOW,void,FILTER> &kwd,
+   const T &val = T{} // <== via SFINAE, T != optional
+) {
+   Node &n = add(val);
+   n.name = kwd.name;
+   return n;
+}
+
+
+
+// -----------------------------------------------------------------------------
+// child_t<plain>, *
+// Guaranteed to add something
+// Returns reference to added node
+// -----------------------------------------------------------------------------
+
+// child_t<plain>, plain
+template<
+   class TYPE, allow ALLOW, class CONVERTER, class FILTER,
+   class T = TYPE,
+   class = typename std::enable_if<!detail::is_optional<T>::value>::type,
+   class = typename std::enable_if<std::is_constructible<TYPE,T>::value>::type
+>
+Node &add(
+   const child_t<TYPE,ALLOW,CONVERTER,FILTER> &kwd,
+   const T &val = T{} // <== via SFINAE, T != optional
+) {
+   try {
+      Node &n = add();
+      kwd.converter(TYPE(val),n);
+      n.name = kwd.name;
+      return n;
+   } catch (...) {
+      log::member("Node.add(" + detail::keyname(kwd) + ",value)");
+      throw;
+   }
+}
+
+// child_t<plain>, optional
+// Not available, because adding it couldn't be certain
+
+// child_t<plain>, defaulted
+template<
+   class TYPE, allow ALLOW, class CONVERTER, class FILTER,
+   class T,
+   class = typename std::enable_if<std::is_constructible<TYPE,T>::value>::type
+>
+Node &add(
+   const child_t<TYPE,ALLOW,CONVERTER,FILTER> &kwd,
+   const defaulted<T> &def
+) {
+   return add(kwd, def.value());
+}
+
+
+
+// -----------------------------------------------------------------------------
+// child_t<optional>, *
+// -----------------------------------------------------------------------------
+
+// child_t<optional>, plain
+// Returns Node &
+template<
+   class TYPE, allow ALLOW, class CONVERTER, class FILTER,
+   class T = TYPE,
+   class = typename std::enable_if<std::is_constructible<TYPE,T>::value>::type
+>
+Node &add(
+   const child_t<std::optional<TYPE>,ALLOW,CONVERTER,FILTER> &kwd,
+   const T &val = T{}
+) {
+   return add(TYPE{}/kwd, val);
+}
+
+// child_t<optional>, optional
+// Returns bool: was something added?
+template<
+   class TYPE, allow ALLOW, class CONVERTER, class FILTER,
+   class T,
+   class = typename std::enable_if<std::is_constructible<TYPE,T>::value>::type
+>
+bool add(
```

Comment:
I'm not a big fan of this signature. Sometimes the `add` function returns a `Node&` and sometimes it returns a `bool`. Someone (likely me) is going to be confused and upset that sometimes this does what he/she wants and sometimes it doesn't.

Reply:
This is actually something that I somewhat agonized over as well. Returning a Node reference tends to be the most useful; think, for example, of the so-called Builder Pattern. Where `std::optional` or `GNDStk::Defaulted` is involved, however, it isn't always guaranteed that something will be added - which we believe will be the correct behavior when and where we use those classes in the high-level interface.

An option would be to just use `bool` everywhere; then, `add()` variations that are guaranteed to add the thing, just always return `true`. Of course I could also always return `void`, but that's less useful. Always returning a Node reference, and throwing an exception if that can't be done, would probably not work well in the context of the use of optional and defaulted objects in the high-level interface. Yet another possibility would be a pointer (nullptr if nothing was added), or perhaps a key/value pair, one of them being bool, as some functions in the C++ standard library do with respect to container reads and writes.

Ultimately, I settled on something that's both maximally useful, and for which differences can be explained and justified (and be easy to remember) in the documentation. Given that I'll be starting the documentation next week, how about we see if I can explain it well enough that it makes sense? If not, then I think it's still early enough (read: no users yet) that it can be changed.